test/multigrid/fixed_coarsening_kernels.cpp

```diff
@@ -0,0 +1,216 @@
+        std::shuffle(base_vec.begin(), base_vec.end(),
+                     std::mt19937{std::random_device{}()});
+        for (gko::size_type i = 0; i < num; i++) {
+            coarse_array.get_data()[i] = base_vec[i + num_rows - num];
+        }
+        return coarse_array;
+    }
+
+    std::unique_ptr<Mtx> gen_mtx(int num_rows, int num_cols)
+    {
+        return gko::test::generate_random_matrix<Mtx>(
+            num_rows, num_cols,
+            std::uniform_int_distribution<>(num_cols, num_cols),
+            std::normal_distribution<value_type>(-1.0, 1.0), rand_engine, ref);
+    }
+
+    void initialize_data(gko::size_type coarse_size = 2)
+    {
+        coarse_rows = gen_coarse_array(coarse_size, m);
+        c_dim = coarse_size;
+
+        d_coarse_rows = gko::Array<index_type>(exec);
+        d_coarse_rows = coarse_rows;
+        restrict_op = Csr::create(ref, gko::dim<2>(c_dim, m), c_dim);
+
+        d_restrict_op = Csr::create(exec);
+        d_restrict_op->copy_from(restrict_op.get());
+        auto system_dense = gen_mtx(m, m);
+        system_mtx = Csr::create(ref);
+        system_dense->convert_to(system_mtx.get());
+
+        d_system_mtx = gko::clone(exec, system_mtx);
+    }
+
+    std::shared_ptr<gko::ReferenceExecutor> ref;
+    std::shared_ptr<gko::EXEC_TYPE> exec;
+
+    std::default_random_engine rand_engine;
+
+    gko::Array<index_type> coarse_rows;
+    gko::Array<index_type> d_coarse_rows;
+
+    std::shared_ptr<Csr> restrict_op;
+    std::shared_ptr<Csr> d_restrict_op;
+
+    std::shared_ptr<Csr> system_mtx;
+    std::shared_ptr<Csr> d_system_mtx;
+
+    gko::size_type n;
+    gko::size_type m;
+    gko::size_type c_dim;
+};
+
+
+TEST_F(FixedCoarsening, GenerateMgLevelIsEquivalentToRef)
+{
+    gko::size_type c_size = 34;
+    initialize_data(c_size);
+    auto mg_level_factory =
+        gko::multigrid::FixedCoarsening<value_type, int>::build()
+            .with_coarse_rows(coarse_rows)
+            .with_skip_sorting(true)
+            .on(ref);
+    auto d_mg_level_factory =
+        gko::multigrid::FixedCoarsening<value_type, int>::build()
+            .with_coarse_rows(d_coarse_rows)
+            .with_skip_sorting(true)
+            .on(exec);
+
+    auto mg_level = mg_level_factory->generate(system_mtx);
+    auto d_mg_level = d_mg_level_factory->generate(d_system_mtx);
+
+    GKO_ASSERT_MTX_NEAR(gko::as<Csr>(d_mg_level->get_prolong_op()),
+                        gko::as<Csr>(mg_level->get_prolong_op()),
+                        r<value_type>::value);
+    GKO_ASSERT_MTX_NEAR(gko::as<Csr>(d_mg_level->get_restrict_op()),
+                        gko::as<Csr>(mg_level->get_restrict_op()),
+                        r<value_type>::value);
+    GKO_ASSERT_MTX_NEAR(gko::as<Csr>(d_mg_level->get_coarse_op()),
+                        gko::as<Csr>(mg_level->get_coarse_op()),
+                        r<value_type>::value);
+}
+
+
+TEST_F(FixedCoarsening, GenerateMgLevelIsEquivalentToRefOnUnsortedMatrix)
```

Comment:
Note: SpGEMM generally requires the second matrix to be sorted, so this only works "accidentally" because you compute it as R*(A*P) and not (R*A)*P

Reply:
Good catch! Looks like I forgot to set the flag to false.